### PR TITLE
HAAR-12 API enabling addition of new email domains to list of allowed domains

### PR DESCRIPTION
**Pull request by AlTurner-MOJ**

### CODE REVIEW

File: src/main/resources/application.yml [hunk 1]
```diff
@@ -128,6 +128,11 @@ application:
     generated-password:
       length: 60
 
+  email.domain.exclude:
+    - gmail
+    - yahoo
+    - hotmail
```
Comment (karenmillermoj) [inline src/main/resources/application.yml:134]: do we need outlook aswell?

Comment (AlTurner-MOJ) [reply src/main/resources/application.yml:134]: Agreed, have added

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/EmailDomainController.kt [hunk 2]
```diff
@@ -1,20 +1,59 @@
...
+    val domainDtoList = emailDomains.map { emailDomain -> EmailDomainDto(emailDomain.id.toString(), emailDomain.name) }
+    return ModelAndView("ui/emailDomains", mapOf("emailDomains" to domainDtoList))
   }
 }
 
 data class EmailDomainDto(val id: String, val domain: String)
+
+data class CreateEmailDomainDto(
+  @field:NotBlank(message = "email domain name must be supplied")
+  @field:Pattern(regexp = "^%.*$", message = "email domain name must start with %")
```
Comment (karenmillermoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/resource/EmailDomainController.kt:53]: nice

File: src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/EmailDomainServiceTest.kt [hunk 3]
```diff
@@ -24,4 +33,41 @@ class EmailDomainServiceTest {
...
+
+  @Test
+  fun `shouldNotAddDomainWhenExcluded`() {
+    whenever(emailDomainExclusions.contains(newDomain.name)).thenReturn(true)
+    assertThatThrownBy { service.addDomain(newDomain) }
+      .isInstanceOf(EmailDomainExcludedException::class.java)
+      .hasMessage("Unable to add email domain: ${newDomain.name} to allowed list with reason: domain present in excluded list")
+
+    verify(emailDomainRepository, never()).save(any())
+  }
```
Comment (karenmillermoj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/EmailDomainServiceTest.kt:57]: fyi - there is also a verifyNoMoreInteractions(emailDomainRepository)

File: src/main/resources/db/dev/data/auth/V900_0__oauth_sample_data.sql [hunk 4]
```diff
@@ -417,7 +418,8 @@ INSERT INTO user_token (token, token_type, token_expiry, user_id) SELECT 'mfa_to
 INSERT INTO user_token (token, token_type, token_expiry, user_id) SELECT 'mfa_code', 'MFA_CODE', '3031-12-10 08:55:45.0000000', user_id from users where username = 'AUTH_MFA_TOKEN_USER';
 
 INSERT INTO roles (role_id, role_code, role_name, role_description, admin_Type)
-VALUES (newid(), 'OAUTH_VIEW_ONLY_CLIENT', 'View Clients', 'View Client details', 'EXT_ADM');
+VALUES (newid(), 'OAUTH_VIEW_ONLY_CLIENT', 'View Clients', 'View Client details', 'EXT_ADM'),
+       (newid(), 'MAINTAIN_EMAIL_DOMAINS', 'Maintain Email Domains', 'Maintain Email Domains for HMPPS Auth', 'EXT_ADM');
```
Comment (karenmillermoj) [inline src/main/resources/db/dev/data/auth/V900_0__oauth_sample_data.sql:422]: Have we also got a VIEW_EMAIL_DOMAINS role for a read only version?

Comment (AlTurner-MOJ) [reply src/main/resources/db/dev/data/auth/V900_0__oauth_sample_data.sql:422]: Hi Karen - we have gone with one role for all access to email domains. Guess if that's an issue we could amend via a change request later?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/EmailDomainService.kt [hunk 5]
```diff
@@ -4,10 +4,32 @@ import org.springframework.stereotype.Service
...
+
+  @Throws(EmailDomainExcludedException::class)
+  fun addDomain(newDomain: CreateEmailDomainDto): List<EmailDomain> {
+    val existingDomain = emailDomainRepository.findByName(newDomain.name)
+    if (existingDomain == null) {
+      if (emailDomainExclusions.contains(newDomain.name)) {
+        throw EmailDomainExcludedException(newDomain.name, "domain present in excluded list")
+      }
+
+      emailDomainRepository.save(EmailDomain(name = newDomain.name, description = newDomain.description))
```
[the remark this answers is not in the document]
Comment (AlTurner-MOJ) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/oauth2server/service/EmailDomainService.kt]: Replaced null check conditional with `existingDomain ?: run { ... }`

**Review by karenmillermoj — APPROVED** [review]
lgtm

**Pull request merged**